lib/ddtrace/contrib/rack/rum_injection.rb

```diff
@@ -0,0 +1,273 @@
+            return false if server_cache?(headers)
+
+            # then check max-age
+            if headers[CACHE_CONTROL_HEADER].include?(MAX_AGE)
+              # only not cached if max-age is 0
+              return headers[CACHE_CONTROL_HEADER].include?(MAX_AGE_ZERO)
+            end
+
+            # not cached if marked no-store, no-cache, or private
+            return (headers[CACHE_CONTROL_HEADER].include?(NO_CACHE) ||
+                   headers[CACHE_CONTROL_HEADER].include?(NO_STORE) ||
+                   headers[CACHE_CONTROL_HEADER].include?(PRIVATE))
+          end
+
+          # last check expires
+          if headers.key?(EXPIRES_HEADER) && !headers[EXPIRES_HEADER].nil?
+            # Expires=0 means not cacced
+            return true if headers[EXPIRES_HEADER] == '0'
+          end
+        end
+
+        def get_current_trace_id
+          tracer = configuration[:tracer]
+          span = tracer.active_span
+          span.trace_id if span
+        end
+
+        def generate_updated_html(response, headers, trace_id)
+          concatted_html = concat_html_fragments(response)
+
+          unix_expiry_time = DateTime.now.strftime('%Q').to_i
+
+          meta_injection_point = find_meta_injection_point(concatted_html)
+          html_comment = html_comment_template(trace_id, unix_expiry_time)
+          meta_tag = meta_tag_template(trace_id, unix_expiry_time) if META_TAG_ENABLED
+
+          modify_html(concatted_html, html_comment, meta_injection_point, meta_tag)
+        # catch everything and swallow it here for defensiveness
+        rescue Exception => e # rubocop:disable Lint/RescueException
+          Datadog.logger.warn("Error updating html for rum injection #{e.class}: #{e.message} #{e.backtrace.join("\n")}")
+          return nil
+        end
+
+        def concat_html_fragments(response)
+          # aggregate the html into a complete document
+          # should have a max amount we parse here after which we give up
+          html_doc = nil
+          response.each do |frag|
+            html_doc ? (html_doc << frag.to_s) : (html_doc = frag.to_s)
+          end
+          # https://www.rubydoc.info/github/rack/rack/file/SPEC
+          # according to spec, .close must be called after iterating
+          response.close if response.respond_to?(:close)
+          html_doc
+        end
+
+        def find_meta_injection_point(html)
+          # we insert directly after head tag open for POC simplicity
+          head_start = html.index(HEAD_TAG_OPEN)
+          html.index(TAG_CLOSE, head_start) + 1 if head_start
+        end
+
+        def meta_tag_template(trace_id, unix_expiry_time)
+          %(<meta name="dd-trace-id" content="#{trace_id}" /> <meta name="dd-trace-time" content="#{unix_expiry_time}" />)
+        end
+
+        def html_comment_template(trace_id, unix_expiry_time)
+          %(<!-- DATADOG;trace-id=#{trace_id};trace-time=#{unix_expiry_time} -->)
```

Comment:
Just a curiosity: why is the current time called `unix_expiry_time` here?

Reply:
just bad naming, can update

Reply:
updated